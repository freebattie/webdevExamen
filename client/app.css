*{
    margin: 0;
    padding: 0;
    border: 0;
    box-sizing: border-box;

    font-family: 'Fira Sans', sans-serif;
}

.menu{
    margin-top: 40px;
    padding: 100px;
    display: flex;

    border: 1px solid black;
    flex-wrap: wrap;
    justify-content: center;

}
.dish{
    margin: 10px;
    width: 300px;
    border: solid 2px black;
    height: 200px;
}
footer{
    width: 80px;
}
.editItem{
    display: flex;
    width: 300px;
    flex-wrap: wrap;
}
.button{
    color: black;
    padding: 5px 2px;
    text-align: center;
    text-decoration: none;
    display: inline-block;
    cursor:pointer;
    margin: 10px;
    font-size: 16px;
    background-color: white;

    border: 2px solid black;
    width: 70px;
    border-radius: 5px;
    transition-duration: 0.4s;
}
.menuheader{
    margin-bottom: 20px;
}
.orders{
    display: flex;
    flex-wrap: wrap;
}
.order{

    height: 300px;
    border: 2px black solid;
    width: 300px;
    margin: 10px;
}
.items{
    width: 100px;
    margin: 10px;
}
.ordersheader{
    width: 100%;
}
.editform{
    width: 400px;
    border: 2px solid black;
}
.itemsToEdit{
    width: 100%;
    display: flex;
    flex-wrap: wrap;

}