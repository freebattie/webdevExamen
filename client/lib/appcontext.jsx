import React from "react";
import { fetchJSON, postJSON } from "./http.js";

export const Appcontext = React.createContext({
  async listDishes() {
    return await fetchJSON("/api/dish");
  },

  async fetchLogin() {
    return await fetchJSON("/api/login");
  },
  async logInUser(user) {
    return await postJSON("/api/login/", user);
  },
  async createUser(user) {
    return await postJSON("/api/login/new", user);
  },
  async addOrder(order) {
    return await postJSON("/api/dish/orders", order);
  },
  async listOrders() {
    return await fetchJSON("/api/dish/orders");
  },
});
