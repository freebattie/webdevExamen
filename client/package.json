{
  "name": "client",
  "version": "1.0.0",
  "description": "",
  "type": "module",
  "scripts": {
    "test": "jest --coverage",
    "format": "prettier --write *.{jsx,js,tsx,ts,html}",
    "dev": "parcel index.html",
    "build": "npm install --include=dev && npm run build:parcel",
    "build:parcel": "parcel build index.html"
  },

  "keywords": [],
  "author": "",
  "license": "ISC",
  "devDependencies": {
    "@babel/preset-env": "^7.19.1",
    "@babel/preset-react": "7.18.6",
    "@types/react": "^18.0.21",
    "@types/react-dom": "^18.0.6",
    "babel-jest": "^29.0.3",
    "events": "^3.3.0",
    "jest": "^29.0.3",
    "jest-environment-jsdom": "^29.0.3",
    "parcel": "^2.7.0",
    "process": "^0.11.10",
    "rimraf": "^3.0.2"
  },
  "dependencies": {
    "@parcel/transformer-react-refresh-wrap": "^2.7.0",
    "@types/jest": "^29.0.3",
    "enzyme": "^3.11.0",
    "history": "^5.3.0",
    "react": "^18.2.0",
    "react-dom": "^18.2.0",
    "react-router-dom": "^6.4.0"
  },
  "jest": {
    "testEnvironment": "jsdom",
    "collectCoverageFrom": [
      "**/*.jsx",
      "**/*.js",
      "!dist/**",
      "!coverage/**",
      "!app.jsx",
      "!index.jsx",
      "!lib/**"
    ]
  },
  "babel": {
    "presets": [
      "@babel/preset-env",
      [
        "@babel/preset-react",
        {
          "runtime": "automatic"
        }
      ]
    ]
  }
}
