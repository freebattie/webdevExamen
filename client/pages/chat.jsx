import { useEffect, useState } from "react";
import { ChatApplication } from "./chatApp.jsx";
const initialMessages = [];

function UserRegistrationForm({ onUsername }) {
  const [username, setUsername] = useState("");
  function handleSubmit(event) {
    event.preventDefault();
    onUsername(username);
  }
  return (
    <center>
      <form className={"loginChat"} onSubmit={handleSubmit}>
        <div className={"loginChatfld"}>
          <label>
            Username:
            <input
              className={"input"}
              value={username}
              onChange={(e) => setUsername(e.target.value)}
            />
          </label>
          <button className={"button"}>Submit</button>
        </div>
      </form>
    </center>
  );
}
export function Chat() {
  const [user, setUser] = useState();
  const [messages, setMessages] = useState(initialMessages);
  const [ws, setWs] = useState();

  useEffect(() => {
    const ws = new WebSocket("ws://" + window.location.host + "/api/chat");
    ws.onopen = (event) => {
      console.log("Opened", event);
    };
    ws.onmessage = (event) => {
      console.log("message", event);
      const { user, message } = JSON.parse(event.data);
      setMessages((messages) => [...messages, { message, user }]);
    };
    setWs(ws);
  }, []);

  function handleNewMessage(message) {
    ws.send(JSON.stringify({ message, user }));
  }

  if (!user) {
    return <UserRegistrationForm onUsername={setUser} />;
  }

  return (
    <ChatApplication messages={messages} onNewMessage={handleNewMessage} />
  );
}
