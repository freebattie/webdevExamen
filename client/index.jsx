import { createRoot } from "react-dom/client";

import { App } from "./app";

// get the root element in index.html
const container = document.getElementById("app");
//creat a react root from that id
const root = createRoot(container);

root.render(<App />);
