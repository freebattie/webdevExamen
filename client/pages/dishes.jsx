import { useContext, useState } from "react";
import { Appcontext } from "../lib/appcontext.jsx";
import { useNavigate } from "react-router-dom";
import { useLoader } from "../lib/useLoading.jsx";
import { ErrorMsg } from "./errormsg.jsx";
import { Dish } from "./dish.jsx";
import { addGivenOrder } from "../lib/addgivenorder.jsx";

export function Dishes({ setError }) {
  const { listDishes } = useContext(Appcontext);
  const [orders, setOrders] = useState([{ price: 0, size: 0 }]);
  const [time, setTime] = useState("08:00");
  const [date, setDate] = useState("");
  const [location, setLocation] = useState("");
  const navigate = useNavigate();
  const { loading, error, data, reload } = useLoader(
    async () => await listDishes()
  );
  const { addOrder } = useContext(Appcontext);

  async function handelAddOrder(e) {
    e.preventDefault();

    const val = [...orders];
    console.log(val);
    let max = 0;
    val.forEach((e) => {
      if (e != null) {
        max += e.price * e.size;
      } else {
      }
    });

    const newOrder = {
      location,
      date: date.length < 1 ? today : date,
      time,
      orders,
      total: max,
    };
    await addGivenOrder(newOrder, setError, navigate, addOrder);
  }
  if (error) {
    return <ErrorMsg error={error} />;
  }
  if (loading) {
    return <h1>Loading....</h1>;
  }
  function getDate() {
    let today = new Date();
    let dd = today.getDate();
    let mm = today.getMonth() + 1; //January is 0!
    let yyyy = today.getFullYear();

    if (dd < 10) {
      dd = "0" + dd;
    }

    if (mm < 10) {
      mm = "0" + mm;
    }

    return yyyy + "-" + mm + "-" + dd;
  }

  let today = getDate();

  return (
    <center>
      <div className={"menuheader"}>DISHES:</div>
      <div className={"menu"}>
        {data.map((d, index) => {
          return (
            <Dish
              key={d.id}
              d={d}
              setOrders={setOrders}
              orders={orders}
              val={0}
              index={index}
            />
          );
        })}
      </div>
      <form onSubmit={(e) => handelAddOrder(e)}>
        <div>
          <label>
            Location:
            <input
              className={"input"}
              type="text"
              value={location}
              onChange={(e) => setLocation(e.target.value)}
            />
          </label>
          <label>
            date:
            <input
              className={"input"}
              type="date"
              value={date}
              min={today}
              onChange={(e) => setDate(e.target.value)}
            />
          </label>
          <label>
            time:
            <input
              className={"input"}
              type="time"
              value={time}
              min={"08:00"}
              max={"21:00"}
              onChange={(e) => setTime(e.target.value)}
            />
          </label>
        </div>
        <button data-testid={"button"} className={"button"}>
          Order
        </button>
      </form>
    </center>
  );
}
