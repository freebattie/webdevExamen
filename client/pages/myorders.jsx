import { useContext } from "react";
import { Appcontext } from "../lib/appcontext.jsx";
import { useLoader } from "../lib/useLoading.jsx";
import { ErrorMsg } from "./errormsg.jsx";
import { useNavigate } from "react-router-dom";
import { LogOut } from "../Components/logout";

export function MyOrders() {
  const { listOrders } = useContext(Appcontext);
  const navigate = useNavigate();
  const { loading, error, data, reload } = useLoader(
    async () => await listOrders()
  );

  if (loading) {
    return <div>Loading...</div>;
  }
  if (error) {
    return <ErrorMsg error={error} />;
  }
  return (
    <div className={"orders"}>
      <div className={"ordersheader"}>
        <h1>ORDER FOR USER: </h1>
        <button className={"button"} onClick={() => navigate("/")}>
          back
        </button>
        <LogOut />
      </div>
      {data.map((o) => {
        return (
          <div key={o.id} className={"order"}>
            <div>name: {o.id}</div>
            <div>date: {o.date}</div>
            <div>time: {o.time}</div>
            <div>location: {o.location}</div>

            {o.orders.map((i, index) => {
              return (
                <div key={index} className={"items"}>
                  <div>name: {i.name}</div>
                  <div>price: {i.price}</div>
                  <div>how many: {i.size}</div>
                </div>
              );
            })}
            <div>total: {o.total}</div>
          </div>
        );
      })}
    </div>
  );
}
