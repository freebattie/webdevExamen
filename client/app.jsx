import { BrowserRouter, Routes, Route } from "react-router-dom";
import { FrontPage } from "./pages/frontpage.jsx";
import { EditUsers } from "./pages/edituser.jsx";
import { ErrorUser } from "./pages/erroruser.jsx";
import { Login } from "./pages/login.jsx";
import { NewUser } from "./pages/newuser.jsx";

import { useContext, useEffect, useState } from "react";
import { MyOrders } from "./pages/myorders.jsx";
import { Error } from "./pages/error.jsx";
import { EditMenu } from "./pages/editmenu.jsx";

export function App() {
  const [error, setError] = useState();
  return (
    <BrowserRouter>
      <Routes>
        <Route path={"/"} element={<FrontPage setError={setError} />} />
        <Route path={"/login"} element={<Login setError={setError} />} />
        <Route path={"/register"} element={<NewUser setError={setError} />} />
        <Route path={"/edit"} element={<EditUsers />} />
        <Route path={"/error"} element={<Error error={error} />} />
        <Route path={"/errorUser"} element={<ErrorUser />} />
        <Route path={"/Myorders"} element={<MyOrders />} />
        <Route path={"/editmenu"} element={<EditMenu />} />
      </Routes>
    </BrowserRouter>
  );
}
