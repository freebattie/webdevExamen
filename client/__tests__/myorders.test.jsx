import * as React from "react";

import { createRoot } from "react-dom/client";
import { act, Simulate } from "react-dom/test-utils";
import { MemoryRouter } from "react-router-dom";
import { MyOrders } from "../pages/myorders";
import { Appcontext } from "../lib/appcontext.jsx";
import { Dishes } from "../pages/dishes.jsx";
global.IS_REACT_ACT_ENVIRONMENT = true;

describe("app", () => {
  it("should rener App", async function () {
    const element = document.createElement("div");

    //creat a React root from that id
    const root = createRoot(element);
    const fn = jest.fn();
    await act(async () => {
      root.render(
        <MemoryRouter>
          <MyOrders />
        </MemoryRouter>
      );
    });
    expect(element).toMatchSnapshot();
    await act(async () => {});
  });

  it("should rener orders", async function () {
    const element = document.createElement("div");

    //creat a React root from that id
    const root = createRoot(element);
    const listDishes2 = [
      {
        id: 0,
        name: "test",
        price: 200,
        description: "d",
        vegan: true,
        type: "pizza",
        orders: [
          {
            name: "test",
            price: 200,
            description: "d",
            vegan: true,
            type: "pizza",
          },
        ],
      },
    ];
    const setError = jest.fn();
    await act(async () => {
      root.render(
        <MemoryRouter initialEntries={["/"]}>
          <Appcontext.Provider value={{ listOrders: () => listDishes2 }}>
            <MyOrders setError={setError} />
          </Appcontext.Provider>
        </MemoryRouter>
      );
    });
    const mockFetchPromise = jest.fn();
    await act(async () => {
      global.fetch = jest.fn().mockImplementation(() => mockFetchPromise);
    });

    expect(element).toMatchSnapshot();
    global.fetch.mockClear();
    delete global.fetch;
  });
});
