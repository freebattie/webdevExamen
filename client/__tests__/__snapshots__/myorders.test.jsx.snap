// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app should rener App 1`] = `
<div>
  <div
    style="border: 1px solid red; background: Pink;"
  >
    An error occurred: 
    ReferenceError: fetch is not defined
  </div>
</div>
`;

exports[`app should rener orders 1`] = `
<div>
  <div
    class="orders"
  >
    <div
      class="ordersheader"
    >
      <h1>
        ORDER FOR USER: 
      </h1>
    </div>
    <div
      class="order"
    >
      <div>
        name: 
        0
      </div>
      <div>
        date: 
      </div>
      <div>
        time: 
      </div>
      <div>
        location: 
      </div>
      <div
        class="items"
      >
        <div>
          name: 
          test
        </div>
        <div>
          price: 
          200
        </div>
        <div>
          how many: 
        </div>
      </div>
      <div>
        total: 
      </div>
    </div>
    <button>
      back
    </button>
    <button>
      back
    </button>
  </div>
</div>
`;
